Fixed issue with registered filename for built debs

hwittenborn opened the pull request:
Closes #111.

lpuv committed:
Switch CI Images for speed

hwittenborn committed:
Bumped version

hwittenborn committed:
Fixed issue with registered filename for built debs

Closes #111.

hwittenborn committed:
Merge branch 'alpha' into version-fix

gamer4life1 commented:
Unit tests are failing :(

hwittenborn commented:
Still trying to figure it out :(. I'm gonna look into it tomorrow (or maybe today if I have the time), and I'm hoping it'll all be done then.

hwittenborn commented:
Hmm, I'm trying to run it locally in Docker any everything, and it all just works fine. Gonna keep looking into it

hwittenborn committed:
Added setup to run tests locally

hwittenborn committed:
Changed commit hash used in PR runs

All of my local testing is making my fix work fine; I have no clue why GitHub Actions is failing the run. Trying this to see if it fixes it.

hwittenborn commented:
Just added a Docker setup for local testing on the unit tests. I'll probably get that going in a separate PR before I merge this.

hwittenborn committed:
Removed Docker config; Updated unit tests to work with '--allow-downgrades' option

hwittenborn committed:
Removed leftover remnants of Docker setup

hwittenborn commented on .drone/scripts/build-native.sh:
```diff
@@ -11,6 +11,10 @@ conflicts="$(echo "${!conflicts}" | sed 's| |, |g')"
 sed -i "s|\$\${pkgname}|${pkgname}|" debian/control
 sed -i "s|\$\${conflicts}|${conflicts}|" debian/control
 
+# Install needed dependencies.
+needed_deps="$(cat debian/control  | grep '^Build-Depends:' | sed 's|^Build-Depends: ||')"
+sudo apt-get satisfy "${needed_deps}" -y
+
```
@gamer4life1: This makes it so that we don't need to rebuild the CI Docker image whenever we update the list of dependencies for makedeb. This kind of contradicts the purpose of the CI image though, but then we get back to if we should be using an image with the deps preinstalled in the first place. Just curious on your thoughts of the benfits vs downsides here, and what approach you'd be interested in taking.

My main gripe against using the CI image is if a contributor were to update the list of build deps, as they wouldn't be able to update the CI image to reflect their new list of dependencies.

hwittenborn commented on .github/workflows/pr.yml:
```diff
@@ -17,6 +17,8 @@ jobs:
         uses: actions/checkout@v2
         with:
           fetch-depth: 0
+          ref: ${{ github.event.pull_request.head_sha }}
+
```
For some reason the CI fails without this. From what I'm reading my change makes it run against the PR's current commit, while it previously ran against an auto-generated commit that would be present without merging.

I'm not sure if this would cause any issues, but it's the only way I've gotten the CI to pass, and I was fine with it and then debugging everything in the future if issues came up.

hwittenborn commented on test/util/util.bash:
```diff
@@ -47,8 +45,46 @@ remove_function() {
     fi
 }
+
+# APT wrapper for common options that we need.
+apt_path="$(type -pf apt-get)"
+
+apt() {
+    args=()
+    cmd="${1}"
+
+    case "${cmd}" in
+        install) args+=('--allow-downgrades') ;;
+    esac
+
+    apt_cmd=("${apt_path}" "${@}" "${args[@]}")
+
+    if [[ -n "${SUDO_PREFIX}" ]]; then
+        sudo "${apt_cmd[@]}"
+    else
+        "${apt_cmd[@]}"
+    fi
+}
+
+apt-get() {
+    apt "${@}"
+}
+
+# Wrap sudo so we can capture it's APT calls.
+sudo() {
+    cmd="${1}"
+
+    case "${cmd}" in
+        apt|apt-get) SUDO_PREFIX=y apt "${@:2}" ;;
+    esac
+}
+
+export -f apt apt-get sudo
+
+# vim: set ts=4 sw=4 expandtab:
```
This makes it so that any calls to `apt install` or `apt-get install` automatically get ran with `--allow-downgrades`, as the CI was failing on some lowered package versions in the CI previously (test packages are recreated throughout the tests and then installed, and sometimes the pkgname's for the packages are the same, which caused the lowered version issue).

hwittenborn reviewed (commented):
Everything should be ready to merge now. I've left a brief note on my new changes below.

gamer4life1 commented:
mmh, LGTM except for the replacement of the ci image. Won't that make it go slow again?

hwittenborn commented:
It would, but I just don't know how to manage deps if users were to add their own from changes in PRs then.

Considering the length of time it takes to run unit tests themselves too, I don't think it really makes much of a difference in the grand scheme of things.

I'll just remove that part for now, and we can discuss a solution later.

gamer4life1 commented:
Sounds good to me.

hwittenborn committed:
Moved dep installation to separate file

hwittenborn committed:
Merged branch 'alpha' into version-fix

hwittenborn committed:
Added Pip dependencies to dependency installation file; Updated GitHub Actions to use dependency installation script

hwittenborn committed:
Moved dependency installation script execution to after repository is cloned in GitHub Actions config

hwittenborn committed:
Fixed syntax errors in GitHub Actions config

hwittenborn committed:
Debugging GitHub Actions changes

hwittenborn committed:
More debugging

hwittenborn committed:
Fixed directory permission issue in GitHub actions

Looks like GitHub actions doesn't throw us into the home directory by default, so our calls with Pip failed previously as it tries to install to the home directory, which is owned by root, and thus causing issues since the makedeb images use the 'makedeb' user by default.

hwittenborn committed:
Added Git installation step before repo cloning so Git repository is properly configured

hwittenborn commented on .drone/scripts/install-deps.sh:
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/bash
+set -ex
+
+# We use a file to manage dependencies instead of in the Drone CI config as it
+# allows us to install dependencies on other platforms such as GitHub Actions.
+#
+# Additionally, we use a single file to install deps instead of one for each CI
+# script as it makes things easier to manage.
+sudo apt-get install asciidoctor bats curl debhelper git gpg jq openssh-client python3 python3-pip python3-requests sed tzdata ubuntu-dev-tools -y
+curl -Ls "https://shlink.${hw_url}/ci-utils" | sudo bash -
+
+pip install PyGithub
```
Moved Python dep installation to this script

hwittenborn committed:
Removed lingering environment key in GitHub Actions config

lpuv reviewed (approved):
LGTM
